v2/src/agent/agent.rs

```diff
@@ -238,21 +238,25 @@ impl Agent {
         let strm = async_stream::try_stream! {
             let docs = Self::get_docs(&messages, &knowledge).await?;
             let tool_descs = Self::get_tool_descs(&tools);
-            loop {
-                let mut model = self.lm.clone();
-                let assistant_out = model.infer(messages.clone(), tool_descs.clone(), docs.clone(), config.clone().unwrap_or_default()).await?;
-                let assistant_msg = assistant_out.message.clone();
-                messages.push(assistant_msg.clone());
-                yield assistant_out;
-
-                if let Some(tool_calls) = assistant_msg.tool_calls && !tool_calls.is_empty() {
-                    for delta in Self::handle_tool_calls(&tools, tool_calls).await? {
-                        let message_output = MessageOutput { message: delta.finish()?, finish_reason: FinishReason::Stop{} };
-                        yield message_output.clone();
-                        messages.push(message_output.message);
+            let mut model = self.lm.clone();
+            let mut need_infer = true;
+            while need_infer {
+                need_infer = false;
+                let mut assistant_strm = model.infer(messages.clone(), tool_descs.clone(), docs.clone(), config.clone().unwrap_or_default());
+                while let Some(assistant_out) = assistant_strm.next().await {
+                    let assistant_out = assistant_out?;
+                    let assistant_msg = assistant_out.message.clone();
+                    yield assistant_out;
+                    messages.push(assistant_msg.clone());
+
+                    if let Some(tool_calls) = assistant_msg.tool_calls && !tool_calls.is_empty() {
+                        for delta in Self::handle_tool_calls(&tools, tool_calls).await? {
+                            let message_output = MessageOutput { message: delta.finish()?, finish_reason: FinishReason::Stop{} };
+                            yield message_output.clone();
+                            messages.push(message_output.message);
+                            need_infer = true;
+                        }
```

Comment:
It seems that `handle_tool_calls` iterates the provided `tool_calls`, so if it's empty the inner for loop is not iterated ever. Hence, I think there's no need to check for `!tool_calls.is_empty()`.
```suggestion
                    if let Some(tool_calls) = assistant_msg.tool_calls {
                        for delta in Self::handle_tool_calls(&tools, tool_calls).await? {
                            let message_output = MessageOutput { message: delta.finish()?, finish_reason: FinishReason::Stop{} };
                            yield message_output.clone();
                            messages.push(message_output.message);
                            need_infer = true;
                        }
```